CIPs/cip-127.md

```diff
@@ -0,0 +1,207 @@
+
+Currently in Ceramic the main accounts types are PKH DID. These are great because they enable existing wallets to be used directly with Ceramic. Once a PKH DID is used to delegate permissions to a session key. That delegation will remain valid until the capability expires. This could be a problem in the case of a stolen session key or a malicious application. By introducing a capability registry these capabilities can be revoked at any time by the main DID (PKH DID in the case above, but this would work for any DID method). A system that uses these object capabilities could refer to the registry to verify that the capability  has not been revoked before it was used.
+
+## Specification
+
+<!--The technical specification should describe the syntax and semantics of any new feature.-->
+This specification describes the data structure of the capability registry, its validation and consensus logic, as well as how wallet UX would look like for someone using the registry.
+
+
+### Capability Registry
+
+The capability registry is based on a self-certifying data structure represented as a special type of Ceramic stream. Each DID has uniquely *one* capability registry. Any object capability issued by this DID can be notarized and revoked in the registry. Every update to the registry is recorded as a *Data Event*.
+
+```verilog
+type Prinicipal Bytes // multidid
+type CACAOHash Bytes // multihash
+type Varsig Bytes // varsig
+
+type Entry struct {
+  key CACAOHash 
+  revoked Boolean // true if the capability has been revoked
+  ocap optional Link // optionally include the CID of the capability
+} representation tuple
+// The registry should eventually be a HAMT or verkle-tree data structure
+type Regsitry { CACAOHash : Entry }
+type Snapshot struct {
+  registry &Regsitry
+  actions [Entry]
+}
+```
+
+#### Ceramic Stream
+
+```verilog
+type Event InitEvent | DataEvent | TimeEvent
+
+type InitEvent &Prinicipal // an inline CID containing raw principal bytes
+
+type DataEvent struct {
+  id &InitEvent
+  prev [&Event] // optional CID pointer to previous event
+  prf [&CACAO] // capabilities used to emit this event
+  data &Snapshot
+  sig Varsig
+}
+
+type EthereumTx // <https://ipld.io/specs/codecs/dag-eth/chain/#transaction-ipld>
+
+type BlockchainTimestamp struct { // https://chainagnostic.org/CAIPs/caip-168
+  root Link
+  chainID String
+  txType String
+  txHash &EthereumTx
+}
+
+type TimeEvent struct {
+  id &InitEvent
+  prev [&DataEvent] // should always be one CID
+  proof &BlockchainTimestamp
+  path String
+}
+```
+
+#### Streamid
+
+Generating the streamid can be done in three steps,
+
+1. Generate the multidid representation of the DID (see [Multidid specification]()).
+
+2. Encode the multidid as an inline CID,
```

Comment:
this is interesting as it breaks some assumptions in the Ceramic code as it exists today.  Currently we always assume the genesis CID can be loaded and used to create an initial StreamState object for the stream